[RFC] Config Override (#3553)
flyteorg/flyte: rfc/system/3553-config-override.md

# Flyte Config Overrides RFC

https://github.com/flyteorg/flyte/issues/475

## Motivation
As a user of Flyte, configuration like "resources (cpu/mem/gpu)", "catalog (enable/disable/version)", "retries", "spark config", "hive cluster" are created only at registration time. The only way to update is to re-register. This is not desirable, as there are times when the user may want to update these parameters for specific launchplans, or executions.

> Using with_overrides can provide node-level override for workflow, but this is still at compile time. Users need to use dynamic workflow if they want to pass the override values as inputs to the workflow. Moreover, with_overrides can not be applied to reference workflow because the workflow function body is not exposed.

Reviewer: ```suggestion
Using `with_overrides` can provide node-level override for workflow, but this is still at compile time. Users need to use dynamic workflow if they want to pass the override values as inputs to the workflow. Moreover, with_overrides can not be applied to reference workflow because the workflow function body is not exposed.
```

Minor nit

## Design

> By default, all task nodes are not overridable at runtime. However, users can call `.with_runtime_override(name: str)` on task nodes to register a "hook" on tasks, which allows the task to be indentified by `name` and overridden with new config at runtime.

Reviewer 2: typo: indentified -> identified.

```python
@task
def t1():
  ...

@task
def t2():
  ...

@workflow
def wf():
  t1() # this task node cannot be overridden

>   t1().with_runtime_override("task-yee") # can be overriden under the name "task-yee"

Reviewer 2: Can we clarify what happens if `t1` defines a pod template? What takes precedence?

Reviewer: @davidmirror-ops let's discuss this point in tomorrows contributors' sync?

  t2().with_runtime_override("task-ketan") # can be overriden under the name "task-ketan"
  t3() # this task node cannot be overridden

> ```

Reviewer: ```suggestion
```

One main motivation for introducing such a "hook mechanism" based on identifiers/names is that we couldn't come up with a good answer how a good UX can be achieved when having to specify overrides *in a nested workflow graph* in the UI, in code, or via the CLI. This way, the overrides can be specified in a simple map and the structure of the workflow graph does not matter/does not have to be shown.
```

We will have a proto called `taskNodeConfigOverride`, which contains all the overridable fields.

```proto

> message TaskNodeConfigOverride {

Reviewer: Does this duplicate logic that already exists for `with_overrides()`? I think it would be nice if these were handled "as similarly as possible".

Author: Will use the original one instead

    bool cache = 1;
    bool cache_serialize = 2;
    string cache_version = 3;

    int32 retries = 4;
    bool interruptible = 5;
    string proxy_as = 6;

    string container_image = 7;
    message ResourcesOverride {
        int64 cpu = 1;
        int64 memory = 2;
        int64 gpu = 3;
    }
    ResourcesOverride requests = 8;
    ResourcesOverride limits = 9;

    map<string, string> environment = 10;
    map<string, string> task_config = 11;
}
```

We provide multiple ways to override the task nodes at runtime, including UI, workflow decorator, launch plan, etc.

### 1. UI

The registered task nodes will prompt on UI for users to assign config override values.

<img src="../images/config-override-ui.png" style="width: 50%; height: 50%">

> ### 2. workflow decorator

Reviewer: @ByronHsu could you please update this with the latest change? 🙏

Users can predefine the override values inside workflow decorator. The value will become the default values on UI.

```python
@workflow(
  runtime_override={
    "task-yee": TaskNodeConfigOverride(...),
    "task-ketan": TaskNodeConfigOverride(...)
  }
)
def wf():
  t1() # this task node cannot be overridden
  t1().with_runtime_override("task-yee") # can be overriden under the name "task-yee"
  t2().with_runtime_override("task-ketan") # can be overriden under the name "task-ketan"
  t3() # this task node cannot be overridden
```

### 3. launch plan

> Users can also provide predefined values by `LaunchPlan.get_or_create`. If users provide both workflow decorator and launch plan predefined values, launch plan one will override workflow decorator one.

Reviewer: ```suggestion
Users can also provide overrides to `LaunchPlan.get_or_create`. If users provide both workflow decorator and launch plan predefined values, launch plan one will override workflow decorator one.
```

```python
launch_plan.LaunchPlan.get_or_create(
    workflow=wf, 
    name="your_lp_name_5", 
    runtime_override={
      "task-yee": TaskNodeConfigOverride(...),
      "task-ketan": TaskNodeConfigOverride(...)
    }
)
```

## Concerns

> 1. How to resolve when two hooks have the same names?

Reviewer 3: both will allow the same override?

Author: we can do that

> - The flytekit compiler should error out

Reviewer 3: it should be the backend compiler.
